src/main/java/org/testng/JarFileUtils.java

```diff
@@ -52,23 +59,41 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
         boolean foundTestngXml = false;
         try (JarFile jf = new JarFile(jarFile)) {
             Enumeration<JarEntry> entries = jf.entries();
+            File file = java.nio.file.Files.createTempDirectory("testngXmlPathInJar-").toFile();
+            String suitePath = null;
             while (entries.hasMoreElements()) {
                 JarEntry je = entries.nextElement();
-                if (matchesXmlPathInJar(je)) {
-                    Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
-                    for (XmlSuite suite : parsedSuites) {
-                        // If test names were specified, only run these test names
-                        XmlSuite suiteToAdd = suite;
-                        if (testNames != null) {
-                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
-                        }
-                        suites.add(suiteToAdd);
+                String jeName = je.getName();
+                if (Parser.canParse(jeName.toLowerCase())){ 
+                    InputStream  inputStream = jf.getInputStream(je);
+                    File copyFile = new File(file, jeName);
+                    Files.copyFile(inputStream, copyFile);
+                    if (matchesXmlPathInJar(je)) {
+                        suitePath = copyFile.toString();
                     }
-                    return true;
                 } else if (isJavaClass(je)) {
                     classes.add(constructClassName(je));
                 }
             }
+            if (Strings.isNullOrEmpty(suitePath)) {
+                return foundTestngXml;
+            }
+            Collection<XmlSuite> parsedSuites = Parser.parse(suitePath,processor); 
+            for (XmlSuite suite : parsedSuites) {
+                // If test names were specified, only run these test names
+                if (testNames != null) {
+                  XmlSuiteUtils xmlSuiteUtils = new XmlSuiteUtils();
+                  xmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
+                  List<String> missMatchedTestname = xmlSuiteUtils.getMissMatchedTestNames(testNames);
```

Comment:
Something is wrong with this method, it won't work if `cloneIfContainsTestsWithNamesMatchingAny` is not called before. Maybe you should pass suite and testNames as constructor parameters.

Reply:
No, `xmlSuiteUtils.getMissMatchedTestNames(testNames)` will still work. Because default value of `matchedTestNames` is a empty list.

Reply:
Yes, but what will be the meaning of the behavior if cloneIfContainsTestsWithNamesMatchingAny has side effects.
I still don't catch the logic behind the class.

Reply:
>I still don't catch the logic behind the class.
Two points:
1. 'TestNG.sanityCheck()'
2. provide service for "-testnames" to find all the matched test names from xml suite. (can be together used with "xmlPathInJar" )